File: LeetCode/Patterns/Backtracking/17. Letter Combinations of a Phone Number/Solution.py
class Solution:
    def letterCombinations(self, digits: str) -> list[str]:
        res = []
        
        # 1. Equivalent to your String[][] arr setup
        arr = [["2","abc"],["3","def"],["4","ghi"],["5","jkl"],["6","mno"],["7","pqrs"],["8","tuv"],["9","wxyz"]]
        
        # 2. Build the dictionary (equivalent to Map<String, String>)
        mapping = {}
        for pair in arr:
            mapping[pair[0]] = pair[1]
            
        # 3. Defining the backtracking helper function
        def backTrack(i, curStr):
            # Base case: if current string length matches digits length
            if len(curStr) == len(digits):
                res.append(curStr)
                return
            
            # Look up the digit (Python handles characters as strings, so no conversion error)
            digit_key = digits[i]
            
            # Loop through characters in the mapped string
            for c in mapping[digit_key]:
                backTrack(i + 1, curStr + c)
                
        # 4. Run only if digits is not empty
        if len(digits) != 0:
            backTrack(0, "")
            
        return res

File: LeetCode/Patterns/Backtracking/17. Letter Combinations of a Phone Number/Solution.java
class Solution {
    public List<String> letterCombinations(String digits) {
        ArrayList<String> res = new ArrayList<>();
        String[][] arr = {{"2","abc"},{"3","def"},{"4","ghi"},{"5","jkl"},{"6","mno"},{"7","pqrs"},{"8","tuv"},{"9","wxyz"}};
        Map<String,String> map = new HashMap<>();
        for(int i=0;i<arr.length;i++){
            map.put(arr[i][0],arr[i][1]);
        }
        
        if (digits.length() != 0){
            backTrack(0,"",digits, res,map);

        }
        return res;
    }
    static void backTrack(int i,String curStr,String digits, List<String> res,Map<String,String> map){
            if(curStr.length() == digits.length()){
                res.add(curStr);
                return;
            }
            String value = String.valueOf(digits.charAt(i));
            for(char c : map.get(value).toCharArray()){
                backTrack(i+1, curStr+c, digits , res,map);
            }
        }
}